subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,1023 @@
+		return -ENODEV;
+	}
+
+	switch (entity.subtype) {
+	case USB_AUDIO_FEATURE_UNIT:
+		return handle_feature_unit_req(audio_dev_data,
+					       pSetup, len, data);
+	case USB_AUDIO_INPUT_TERMINAL:
+	case USB_AUDIO_OUTPUT_TERMINAL:
+	case USB_AUDIO_MIXER_UNIT:
+	case USB_AUDIO_SELECTOR_UNIT:
+	case USB_AUDIO_PROCESSING_UNIT:
+	case USB_AUDIO_EXTENSION_UNIT:
+	default:
+		LOG_INF("Currently not supported");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Handler called for class specific endpoint request.
+ *
+ * This function handles all class specific endpoint requests to a usb audio
+ * device. If request is properly server then 0 is returned. Returning negative
+ * value will lead to set stall on IN EP0.
+ *
+ * @param pSetup    Information about the executed request.
+ * @param len       Size of the buffer.
+ * @param data      Buffer containing the request result.
+ *
+ * @return  0 on success, negative errno code on fail.
+ */
+static int handle_endpoint_req(struct usb_setup_packet *pSetup,
+				      s32_t *len,
+				      u8_t **data)
+{
+	return -1;
+}
+
+/**
+ * @brief Handler called for Class requests not handled by the USB stack.
+ *
+ * @param pSetup    Information about the request to execute.
+ * @param len       Size of the buffer.
+ * @param data      Buffer containing the request result.
+ *
+ * @return  0 on success, negative errno code on fail.
+ */
+int audio_class_handle_req(struct usb_setup_packet *pSetup,
+			   s32_t *len, u8_t **data)
+{
+	LOG_INF("bmRequestType 0x%02x, bRequest 0x%02x, wValue 0x%04x,"
+		"wIndex 0x%04x, wLength 0x%04x",
+		pSetup->bmRequestType, pSetup->bRequest, pSetup->wValue,
+		pSetup->wIndex, pSetup->wLength);
+
+	switch (REQTYPE_GET_RECIP(pSetup->bmRequestType)) {
+	case REQTYPE_RECIP_INTERFACE:
+		if (handle_interface_req(pSetup, len, data) < 0) {
+			return -1;
+		}
+		break;
+	case REQTYPE_RECIP_ENDPOINT:
+		if (handle_endpoint_req(pSetup, len, data) < 0) {
+			return -1;
+		}
+		break;
+	default:
+		LOG_ERR("Request receipent invalid");
+		return -1;
+	}
+
+	return 0;
+}
+
+static const struct usb_audio_device_api {
+	void (*init)(void);
+} usb_audio_api;
+
+static int usb_audio_device_init(struct device *dev)
+{
+	LOG_DBG("Init Audio Device: dev %p (%s)", dev, dev->config->name);
+
+	return 0;
+}
+
+static void audio_write_cb(u8_t ep, int size, void *priv)
+{
+	struct usb_dev_data *dev_data;
+	struct usb_audio_dev_data_t *audio_dev_data;
+	struct net_buf *buffer = priv;
+
+	dev_data = usb_get_dev_data_by_ep(&usb_audio_data_devlist, ep);
+	audio_dev_data = dev_data->dev->driver_data;
+
+	LOG_DBG("Written %d bytes on ep 0x%02x, *audio_dev_data %p",
+		size, ep, audio_dev_data);
+
+	/* Release net_buf back to the pool */
+	net_buf_unref(buffer);
+
+	/* Call user callback if user registered one */
+	if (audio_dev_data->ops && audio_dev_data->ops->data_written_cb) {
+		audio_dev_data->ops->data_written_cb(dev_data->dev, NULL, size);
```

Comment:
I will reword the API. Same callback type is used for send/received there is where the issue comes from. I will distinguish the behavior when it comes to the buffer.
I think there is no any reason so that the user must free the buffer after send so this will remain NULL.

Reply:
Discussed offline and changed my mind. Will pass the buffer and depending on user app (defined the callback or not). The user is responsible/not responsible for freeing the buffer. This is more consistent approach.